File: 47-permutations-ii/permutations-ii.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    const results = []
    const counter = {}

    for (let num of nums) {
        if (!(num in counter)) counter[num] = 0
        counter[num]++
    }

    console.log(counter)

    const backtrack = (combination) => {
        if (combination.length === nums.length) {
            results.push([...combination])
            return
        }

        for (let num in counter) {
            if (counter[num] === 0) continue
            counter[num]--
            combination.push(parseInt(num))
            backtrack(combination)
            counter[num]++
            combination.pop()
        }

    }

    backtrack([])
    return results
};

File: 47-permutations-ii/permutations-ii.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        
        def backtrack(counter, path):
            if len(path) == len(nums):
                res.append(path.copy())
                return
            for num in counter:
                if counter[num] > 0:
                    counter[num] -= 1
                    path.append(num)
                    backtrack(counter, path)
                    path.pop()
                    counter[num] += 1

        res = []
        backtrack(Counter(nums), [])
        return res